### PR TITLE
Added to_byte method for issue #329

### PR DESCRIPTION
**Pull request by KrisTC**
I have added the to_byte method and updated the unit tests.

This gives slightly nicer syntax than static_cast, is better than the
c-style cast used in the unit test.

See: https://github.com/Microsoft/GSL/issues/329#issuecomment-240588515

I have tested my change on:
* VS2013
* VS2015
* Apple LLVM version 7.3.0 (clang-703.0.31)

Hopefully everything passes

### CODE REVIEW

File: gsl/gsl_byte [hunk 1]
```diff
@@ -106,6 +106,12 @@ constexpr IntegerType to_integer(byte b) noexcept
     return {b};
 }
 
+constexpr byte to_byte(unsigned char i) noexcept
+{
```
Comment (gdr-at-ms) [inline gsl/gsl_byte]: This version allows "accidental" narrowing conversion. Could you cook a version that weed out non-'unsigned char' via sfinae trickery?  E.g. something along the line of
```
template<typename T, typename = std::enable_if_t<std::is_same<T, unsigned char>::value, T>>
constexpr byte to_byte(T t) noexcept
{
     return byte(t);
}
```

Comment (KrisTC) [inline gsl/gsl_byte]: I have confirmed that in Apple LLVM version 7.3.0 (clang-703.0.31) I definitely get accidental narrowing. I am used to C4244 from MS compiler. I tried simply putting your suggestion in, but it fails the unit test on the constants:
```
byte a = to_byte(0x00); // Error with is_same<T, unsigned char>
```
I have had a look at the available traits: http://en.cppreference.com/w/cpp/header/type_traits and I am not seeing anything else that will help.
I played about with overloads static asserts:
```
template <int I>
constexpr byte to_byte() noexcept
{
    static_assert(I >= 0 && I <= 255, "Value must be in range 0-255");
    return static_cast<byte>(I);
}

template <int I>
constexpr byte to_byte_v = to_byte<I>();
```
Which give the following for constants:
```
byte a = to_byte<0x00>();       // a bit odd
byte b = to_byte_v<0x00>;       // maybe better, still nicer than:
byte c = static_cast<byte>(0x00);
```
But I think they are a bit weird. I will think about it some more and see if there is anything I can come up with.

Comment (KrisTC) [inline gsl/gsl_byte]: I also experimented with initializer_list's to get syntax like:
```
byte d = to_byte({0x00});
```
Which I think looks better and works better. 
But allows invalid calls:
```
byte e = to_byte({});
byte f = to_byte({0x00, 0x00});
```
I can't see anyway ensure 1 entry.
@gdr-at-ms: Do options a, b or d look like they might be acceptable? Option b won't work with VS2013 :(

**Comment by KrisTC** [comment]
I ended up going with: `gsl::to_byte<0x00>()` for constants and `gsl::to_byte(some_uchar)` for variables. I didn't keep the `gsl::to_byte_v<0x00>` version as it doesn't work with VS2013 and caused my VS2015 to have an internal compiler error. 
I have a static asserts for cases where the wrong type is passed or the constant is out of range, giving dev instructions on how to use to_byte correctly.

I think this is the best I could come up with, I hope it's ok.

**Comment by neilmacintosh** [comment]
@KrisTC just quickly...if you saw an internal compiler error from VS2015, could you please send us a repro? We take fixing those crashes very seriously and I want to make sure we get that one.

**Comment by KrisTC** [comment]
@neilmacintosh sure, I will do that tomorrow. I still wasn't keen on having the API `#if def'd` out for VS2013. I think it is better to have consistency.

**Comment by KrisTC** [comment]
@neilmacintosh, I went to report the error using VS feedback and it is already logged: "constexpr variable template cause internal compiler error" (last modified 10th Aug) - I up voted it.

**Comment by neilmacintosh** [comment]
@KrisTC perfect, thanks! I agree we don't want APIs that are `#ifdef`d out if at all possible, I just wanted to make sure that Microsoft was tracking the defect in the compiler. Thanks for upvoting that bug.

**Comment by KrisTC** [comment]
@gdr-at-ms is there anything else you would like me to do for this?
Thanks Kris

**Comment by gdr-at-ms** [comment]
The distinction between constants and runtime value is a little bit troubling, but I suspect this is strictly an improvement over the current situation, so it looks good to me.  
@neilmacintosh , would you want me to pull the merge?

**Comment by KrisTC** [comment]
@gdr-at-ms, I agree, I couldn't come up with anything cleaner :(
I assume the proposed changes to the standard will have the same issues as my my methods:

```
std::byte b1 = 0xFFFF; // Compile's ok but narrowing
std::byte b2 = {0xFFFF}; // Compile's fail!
std::byte b3 = 0xFF; // OK
std::byte b4 = {0xFF}; // OK
```

**Comment by gdr-at-ms** [comment]
@KrisTC With the proposed changes to the standard (which are in the CD out for ballot), you need to systematically use the braces, and narrowing will be enforced.  There is no provision to convert from *bare integer* to an enum -- that has always been banned in C++.

**Comment by neilmacintosh** [comment]
LGTM too! Thanks for taking this on @KrisTC and congratulations on your first PR!

**Pull request merged**